Feature nrf52 low power ble

bbugl opened the pull request:
### Summary of changes <!-- Required -->

The nRF Cordio driver doesn't turn off all the peripherals correctly, resulting in a current consumption that is too high. This PR fixes this behaviour.

#### Impact of changes <!-- Optional -->

None.

#### Migration actions required <!-- Optional -->

None.

### Documentation <!-- Required -->

None.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@pan- 
@AGlass0fMilk 
----------------------------------------------------------------------------------------------------------------

bbugl committed:
NRF52: Fix errata_20

Nordic documentation suggests that this needs to be done regardless of which RTC is used.

bbugl committed:
BLE: Remove hardcoded use of external low frequency crystal oscillator

If the internal RC oscillator is used it needs to be calibrated after the HF clock is started.

bbugl committed:
NRF52: Stop RTC and shutdown timer for cordio when not in use

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/common_rtc.c:
```diff
@@ -100,9 +100,6 @@ void COMMON_RTC_IRQ_HANDLER(void)
 __STATIC_INLINE void errata_20(void)
 {
 #if defined(NRF52_PAN_20)
-    if (COMMON_RTC_INSTANCE != NRF_RTC1) {
```
Thanks for addressing this defect in the HAL.

pan- reviewed (dismissed):
Much appreciated @bbugl . I think it can be integrated into the feature branch.

bbugl committed:
NRF52: Add dependency on cordio

bbugl committed:
NRF52: Fix bare metal builds

Bare metal builds are not necessarily dependent on BLE or the Cordio stack, therefore BLE and Cordio related things should only be included when they are used.

bbugl committed:
NRF52: Add pal_rtc.c file to cmake for NRFCordioHCIDriver

bbugl committed:
NRF52: Use COMMON_RTC macros consistently

Add a COMMON_RTC_IRQN macro to abstract RTC instance away completely

bbugl committed:
BLE: Turn off Timer0 completely after Adv for real low power operation

bbugl committed:
NRF52: Revert some changes to hal_sleep

I don't understand why they were made and I couldn't measure any difference.

bbugl committed:
NRF52: Remove redundant (and unterminated) #if

This is a remnant of reverting back to an older version of sleep.
